README
mod_carddav:
------------
-this is an Apache module implementation of CardDAV support to WebDAV
-is dependent on mod_dav_acl <http://sourceforge.net/projects/moddavacl/>,
which implements RFC 3744 also as an Apache module
-uses vcard library of the Evolution Data Server (EDS)

Configuration:
--------------
-must be loaded (dlopened) after mod_dav. Depends on the base library of
mod_dav_acl.
-once the module installs, some directives from the mod_dav_acl are needed:

DAVETagResponse on
-enable strong ETag responses

DAVACL on
-enable ACL handling

CardDAV on
-enables carddav handling in a directory


Install:
--------
-preferably e.g. debuild or rpm-build. Otherwise ./configure && make &&
make install. You may need to run autogen.sh to generate a proper
./configure etc.

in debian:"sudo a2enmod dav_carddav" enables module loading and basic
configuration (though you must enable carddav in appropriate collections).


Tests:
------
-once the module is successfully compiled and loaded, tests can be run with
make check or by running ./carddav_tests. You need to have admin/password
(digest) account and /xxx/carddav directory must be available and accessible by
apache


ChangeLog
Look debian/changelog for later changes

07.07.2008 Jari Urpalainen
  * mod_carddav.c
   -added conditional REPORTs (If-None-Match: "zyx")

06.06.2008 Jari Urpalainen
  * Initial release 0.0.1



vcard_tests.c
/*
 * Test cases for carddav search
 * This is part of a mod_carddav library.
 *
 * Copyright (C) 2011 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include "config.h"

#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>

#include <glib.h>
#include <libebook/e-vcard.h>
#include <glib/gstdio.h>

#include <libxml/tree.h>

#include "mod_carddav.h"
#include "carddav_vcard.h"

#define NS_DAV	"DAV:"


struct state {
	int dump;
} state[1] = {{ 0 }};


static void get_nodes(xmlDoc *doc, xmlNode **s, xmlNode **f)
{
    xmlNode *prop = NULL, *node = doc ? doc->children : NULL,
	       *carddata = NULL, *filter = NULL;

    for ( ; node; node = node->next) {
	if (NODE_NOT_CARDDAV(node)) {
	    ;
	}
	else if (NODE_MATCH(node, "addressbook-query")) {
	    FOR_CHILD(prop, node) {
		if (NODE_NOT_DAV(prop))
		    ;
		else if (NODE_MATCH(prop, "prop"))
		    break;
	    }
	    break;
	}
    }

    FOR_CHILD(carddata, prop) {
	if (NODE_NOT_CARDDAV(carddata))
	    ;
	else if (NODE_MATCH(carddata, "address-data"))
	    break;
    }

    filter = prop ? prop->parent : NULL;

    FOR_CHILD(filter, filter) {
	if (NODE_NOT_CARDDAV(filter))
	    ;
	else if (NODE_MATCH(filter, "filter"))
	    break;
    }

    *s = carddata;
    *f = filter;
}

static int search(const char *tmp, xmlDoc *doc, xmlNode *carddata,
                  xmlNode *filter, int ret)
{
    carddav_search_t *p;
    int rc;

    if ((rc = carddav_vcard_search(tmp, carddata, filter, &p))) {
	char *pch = NULL;

	if (state->dump)
	    printf("%s\n", pch = carddav_vcard_dump(p));
	free(pch);
    }
    else if (ret) {
	fprintf(stderr, "search did not find anything\n");
    }

    carddav_vcard_free(p);

    xmlFreeDoc(doc);
    g_unlink(tmp);

    return rc;
}

static int test(const char *buf, const char *vcard, int ret)
{
    xmlDoc *doc = xmlReadMemory(buf, strlen (buf), NULL, NULL,
				XML_PARSE_NOWARNING);
    xmlNode *carddata = NULL, *filter = NULL;
    char tmp[] = "/tmp/vcard-XXXXXX";

    g_mkstemp(tmp);
    g_assert(g_file_set_contents(tmp, vcard, strlen(vcard), NULL) == TRUE);

    get_nodes(doc, &carddata, &filter);
    g_assert(carddata != NULL);

    g_assert(search(tmp, doc, carddata, filter, ret) == ret);
}

static const char vcard1[] =
	"BEGIN:VCARD\r\n"
	"VERSION:3.0\r\n"
	"FN:Cyrus Daboo\r\n"
	"N:Daboo;Cyrus\r\n"
	"ADR;TYPE=POSTAL:;2822 Email HQ;Suite 2821;RFCVille;PA;15213;USA\r\n"
	"EMAIL;TYPE=INTERNET,PREF:cyrus@example.com\r\n"
	"NICKNAME:foo,me\r\n"
	"NOTE:Example VCard.\r\n"
	"ORG:Self Employed\r\n"
	"TEL;TYPE=WORK,VOICE:412 605 0499\r\n"
	"TEL;TYPE=FAX:412 605 0705\r\n"
	"URL:http://www.example.com\r\n"
	"UID:1234-5678-9000-1\r\n"
	"END:VCARD\r\n";

static const char vcard2[] =
	"BEGIN:VCARD\r\n"
	"VERSION:3.0\r\n"
	"REV:2008-02-18T10:44:26Z\r\n"
	"UID:47B9618A00000004\r\n"
	"TEL;TYPE=CELL:123400002\r\n"
	"N:Pheckc;Jjsccj;;;\r\n"
	"PHOTO;TYPE=\"X-EVOLUTION-UNKNOWN\";ENCODING=b:/9j/4AAQSkZJRgABAQEARwBHAAD//gAXQ3JlYXRlZCB3aXRoIFRoZSBHSU1Q/9sAQwAIBgYHBgUIBwcHCQkICgwUDQwLCwwZEhMPFB0aHx4dGhwcICQuJyAiLCMcHCg3KSwwMTQ0NB8nOT04MjwuMzQy/9sAQwEJCQkMCwwYDQ0YMiEcITIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIy/8AAEQgAMgAyAwEiAAIRAQMRAf/EABsAAQACAwEBAAAAAAAAAAAAAAAHCAQFBgID/8QAMBAAAgEDAQYEBQQDAAAAAAAAAQIDAAQRBQYSEyExQQdhcYEiI0JRkRQVMqFiguH/xAAaAQADAQEBAQAAAAAAAAAAAAAABAUCBgED/8QAIxEAAgICAQQCAwAAAAAAAAAAAAECAwQRQRITITEUYQUiUf/aAAwDAQACEQMRAD8An+sHUtWtNKjVrmQ7754cajLvjrgfbzPIdzWdVfds9pJb3XdQkMrcFZGj+HqY0bdVV9Tz/wBia+N9vbjvkaxMb5E9N6SJB1HxLEEjJaWsUjD6QzSMPXdGB7E1zV74t63HINy1s4F7CWCTn77wrA0TY86jY3N1qsUk6wxBxBDvYjLHkoUH4j3JP/a0V3s1CvF/QM9tKpw0THeU+TLkj8VLnmzT8y0n9FujBx5bioba/rZLWx3iPZ7RzLp95GtnqRGVTezHNjruH7/4n+67iqpq7Qi3uYWMMsNynfnE6sM8/Lr6VamFi0KMepUE1Sx7XZHbI+fjxos1H0z3SlKYEjzISI2I64OKqsyu8sck2QYrmPjBvpIYg598Vauoh8VtlY7JW2isoBwpPl6hGByZTyD+o6E+h7UtlVOcPHA/+PyI1Wal6Zp7vaC/06wnTTLtEeUDiKwzu4H8vI9AM9Tiuctkng1Nnk1G5cOoYifB4nI/jB7VjWuoT21qPmwXUCHKlphHKvqG5N6g0/cLi/Rg88FhbkbxlaUSu3kqpnn6kDzqGqbNdPB0XyK4/svZr9RVntL50GePdcKEDqzhVBx7sKtPpayppNosxzKIlDHzxUFeG2zo2n2kivWhK6PpHwwoTnfk65J7kZyT9z5VYADAwKuYtfRA5zPv7tnjgUpSmREV8bq1hvbWW1uY1khlUo6MMhgeor7UoAje18FtmLe9eeQT3EXPcglkJRPbv71EWu7Dajp2o3MGmlRCkjKQ30jPUe1WlrlNW0RptTleNB84DnjkD0P9VlxT4Nqck9pmn8JuFp2zo0cgCWFi2e7555/NSHXLadso2m3sU0NxlV65HM+VdTW3rgwvsUpSvAFKUoAUxSlAClKUAKUpQB//2Q==\r\n"
	"CALURI:http://example.com/foo/2\r\n"
	"X-EVOLUTION-FILE-AS:Toshok\\, Chris\r\n"
	"FN:Jjsccj Pheckc\r\n"
	"EMAIL;TYPE=INTERNET:toshok@ximian.com\r\n"
	"ORG:Ximian\\, Inc.;\r\n"
	"END:VCARD\r\n";

static void nickname(void)
{
    const char buf[] =
	"<?xml version=\"1.0\" encoding=\"utf-8\" ?>"
	"<C:addressbook-query xmlns:D=\"DAV:\""
	"  xmlns:C=\"urn:ietf:params:xml:ns:carddav\">"
	"  <D:prop>"
	"    <D:getetag/>"
	"    <C:address-data>"
	"      <C:prop name=\"VERSION\"/>"
	"      <C:prop name=\"UID\"/>"
	"      <C:prop name=\"NICKNAME\"/>"
	"      <C:prop name=\"EMAIL\"/>"
	"      <C:prop name=\"FN\"/>"
	"    </C:address-data>"
	"  </D:prop>"
	"  <C:filter>"
	"    <C:prop-filter name=\"NICKNAME\">"
	"      <C:text-match collation=\"i;unicode-casemap\""
	"                    match-type=\"equals\""
	"       >me</C:text-match>"
	"    </C:prop-filter>"
	"  </C:filter>"
	"</C:addressbook-query>";

    test(buf, vcard1, FALSE);
    test(buf, vcard2, FALSE);
}

static void cell(void)
{
    const char buf[] =
	"<?xml version=\"1.0\" encoding=\"utf-8\" ?>"
	"<C:addressbook-query xmlns:D=\"DAV:\""
	"  xmlns:C=\"urn:ietf:params:xml:ns:carddav\">"
	"  <D:prop>"
	"    <D:getetag/>"
	"    <C:address-data>"
	"      <C:prop name=\"VERSION\"/>"
	"      <C:prop name=\"UID\"/>"
	"      <C:prop name=\"NICKNAME\"/>"
	"      <C:prop name=\"EMAIL\"/>"
	"      <C:prop name=\"FN\"/>"
	"      <C:prop name=\"TEL\"/>"
	"    </C:address-data>"
	"  </D:prop>"
	"  <C:filter test=\"allof\">"
	"    <C:prop-filter name=\"TEL\">"
	"       <C:param-filter name=\"TYPE\">"
	"             <C:text-match collation=\"i;unicode-casemap\""
	"                    match-type=\"equals\""
	"       >CELL</C:text-match>"
	"       </C:param-filter>"
	"    </C:prop-filter>"
	"  </C:filter>"
	"</C:addressbook-query>";

    test(buf, vcard1, FALSE);
    test(buf, vcard2, TRUE);
}

static void dual(void)
{
    const char buf[] =
	"<?xml version=\"1.0\" encoding=\"utf-8\" ?>"
	"<C:addressbook-query xmlns:D=\"DAV:\""
	"  xmlns:C=\"urn:ietf:params:xml:ns:carddav\">"
	"  <D:prop>"
	"    <D:getetag/>"
	"    <C:address-data>"
	"      <C:prop name=\"VERSION\"/>"
	"      <C:prop name=\"UID\"/>"
	"      <C:prop name=\"NICKNAME\"/>"
	"      <C:prop name=\"EMAIL\"/>"
	"      <C:prop name=\"FN\"/>"
	"      <C:prop name=\"NOTE\"/>"
	"    </C:address-data>"
	"  </D:prop>"
	"  <C:filter test=\"allof\">"
	"    <C:prop-filter name=\"NICKNAME\">"
	"      <C:text-match collation=\"i;unicode-casemap\""
	"                    match-type=\"starts-with\""
	"       >foo</C:text-match>"
	"    </C:prop-filter>"
	"    <C:prop-filter name=\"NOTE\">"
	"     <C:text-match collation=\"i;unicode-casemap\""
	"                    match-type=\"contains\""
	"       >xample</C:text-match>"
	"    </C:prop-filter>"
	"  </C:filter>"
	"</C:addressbook-query>";

    test(buf, vcard1, TRUE);
    test(buf, vcard2, FALSE);
}

static void param_text(void)
{
    const char buf[] =
	"<?xml version=\"1.0\" encoding=\"utf-8\" ?>"
	"<C:addressbook-query xmlns:D=\"DAV:\""
	"  xmlns:C=\"urn:ietf:params:xml:ns:carddav\">"
	"  <D:prop>"
	"    <D:getetag/>"
	"    <C:address-data>"
	"      <C:prop name=\"VERSION\"/>"
	"      <C:prop name=\"UID\"/>"
	"      <C:prop name=\"NICKNAME\"/>"
	"      <C:prop name=\"EMAIL\"/>"
	"      <C:prop name=\"FN\"/>"
	"      <C:prop name=\"TEL\"/>"
	"    </C:address-data>"
	"  </D:prop>"
	"  <C:filter test=\"allof\">"
	"    <C:prop-filter name=\"TEL\" test=\"allof\">"
	"       <C:param-filter name=\"TYPE\">"
	"             <C:text-match collation=\"i;unicode-casemap\""
	"                    match-type=\"equals\""
	"       >CELL</C:text-match>"
	"              </C:param-filter>"
	"        <C:text-match collation=\"i;unicode-casemap\""
	"                    match-type=\"contains\""
	"       >1234</C:text-match>"
	"    </C:prop-filter>"
	"  </C:filter>"
	"</C:addressbook-query>";

    test(buf, vcard1, FALSE);
    test(buf, vcard2, TRUE);
}

static void anyof(void)
{
    const char buf[] =
	"<?xml version=\"1.0\" encoding=\"utf-8\" ?>"
	"<C:addressbook-query xmlns:D=\"DAV:\""
	"  xmlns:C=\"urn:ietf:params:xml:ns:carddav\">"
	"  <D:prop>"
	"    <D:getetag/>"
	"    <C:address-data>"
	"      <C:prop name=\"VERSION\"/>"
	"      <C:prop name=\"UID\"/>"
	"      <C:prop name=\"NICKNAME\"/>"
	"      <C:prop name=\"EMAIL\"/>"
	"      <C:prop name=\"FN\"/>"
	"      <C:prop name=\"TEL\"/>"
	"    </C:address-data>"
	"  </D:prop>"
	"  <C:filter test=\"allof\">"
	"    <C:prop-filter name=\"TEL\" test=\"anyof\">"
	"       <C:param-filter name=\"TYPE\">"
	"             <C:text-match collation=\"i;unicode-casemap\""
	"                    match-type=\"equals\""
	"       >FAX</C:text-match>"
	"              </C:param-filter>"
	"        <C:text-match collation=\"i;unicode-casemap\""
	"                    match-type=\"contains\""
	"       >1234</C:text-match>"
	"    </C:prop-filter>"
	"  </C:filter>"
	"</C:addressbook-query>";

    test(buf, vcard1, TRUE);
    test(buf, vcard2, TRUE);
}

static void empty(void)
{
    const char buf[] =
	"<?xml version=\"1.0\" encoding=\"utf-8\" ?>"
	"<C:addressbook-query xmlns:D=\"DAV:\""
	"  xmlns:C=\"urn:ietf:params:xml:ns:carddav\">"
	"  <D:prop>"
	"    <D:getetag/>"
	"    <C:address-data/>"
	"  </D:prop>"
	"  <C:filter>"
	"    <C:prop-filter name=\"UID\"/>"
	"  </C:filter>"
	"</C:addressbook-query>";

    test(buf, vcard1, TRUE);
    test(buf, vcard2, TRUE);
}


int main(int argc, char *argv[])
{
    int ret, opt_index;

    static const struct option opt_tbl[] = {
	{ "dump",	no_argument, NULL, 'g' },
	{ "help",	no_argument, NULL, 'h' },
	{ 0 }
    };

    g_type_init();

    g_test_init(&argc, &argv, NULL);

    while ((ret = getopt_long(argc, argv, "gh", opt_tbl, &opt_index)) != -1)
	switch (ret) {
	case 'g':
	    state->dump = TRUE;
	    break;

	case 'h':
	default:
	    printf(
		"Usage: %s [options]\n"
		"    options:\n"
		"       -g (--dump) dump searched content to stdout\n"
		"\n"
		"to see g_test parameter options: run with option -? (--help)\n",
	    argv[0]);
	    return EXIT_SUCCESS;
	}

    g_test_add_func("/testvcard/anyof", anyof);
    g_test_add_func("/testvcard/nickname", nickname);
    g_test_add_func("/testvcard/dual", dual);
    g_test_add_func("/testvcard/cell", cell);
    g_test_add_func("/testvcard/param_text", param_text);
    g_test_add_func("/testvcard/empty", empty);

    return g_test_run();
}


test_carddav.c
/**
 * This is part of a mod_carddav library.
 *
 * Copyright (C) 2011 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include "config.h"

#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <assert.h>
#include <signal.h>
#include <ctype.h>
#include <unistd.h>

#include <glib.h>
#include <libebook/e-vcard.h>

#include <libxml/tree.h>

#define NS_DAV	"DAV:"

#include "mod_carddav.h"
#include "carddav_vcard.h"


static void usage(const char *argv)
{
    fprintf(stdout, "%s -q request_xml_doc -i vcardfile\n" \
		    "returns 0 if succeeds, 1 for an error\n", argv);
}

/** main for vcard search testings */
int main(int argc, char *argv[])
{
    int opt, rc;
    const char *file = NULL, *vcard = NULL;
    xmlDocPtr doc;
    carddav_search_t *p = NULL;
    xmlNodePtr node, carddata = NULL, child = NULL;

    g_type_init();

    while ((opt = getopt(argc, argv, "q:i:")) != -1) {
	switch (opt) {
	case ':':
	case 'h':
	case '?':
	    usage(argv[0]);
	    return EXIT_FAILURE;

	case 'q':
	    file = optarg;
	    break;

	case 'i':
	    vcard = optarg;
	    break;
	}
    }
    if (file == NULL || vcard == NULL) {
	usage(argv[0]);
	return EXIT_FAILURE;
    }
    /* do not prevent the creation of cdata nodes */
    doc = xmlParseFile(file);

    node = doc ? doc->children : NULL;

    for ( ; node; node = node->next) {
	if (NODE_NOT_CARDDAV(node)) {
	    ;
	}
	else if (NODE_MATCH(node, "addressbook-query")) {
	    FOR_CHILD(node, node) {
		if (NODE_NOT_DAV(node))
		    ;
		else if (NODE_MATCH(node, "prop"))
		    break;
	    }
	    break;
	}
    }

    FOR_CHILD(carddata, node) {
	if (NODE_NOT_CARDDAV(carddata))
	    ;
	else if (NODE_MATCH(carddata, "address-data"))
	    break;
    }

    child = node ? node->parent : NULL;
    FOR_CHILD(child, child) {
	if (NODE_MATCH(child, "filter"))
	    break;
    }

    if ((rc = carddav_vcard_search(vcard, carddata, child, &p))) {
	char *pch = NULL;

	printf("%s\n", pch = carddav_vcard_dump(p));
	free(pch);
    }
    else {
	printf("search did not find anything\n");
    }

    carddav_vcard_free(p);

    xmlFreeDoc(doc);

    xmlCleanupParser();

    return rc ? EXIT_SUCCESS : EXIT_FAILURE;
}



mod_carddav.h
/**
 * This is part of a mod_carddav library.
 *
 * Copyright (C) 2008 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */
#ifndef _MOD_CARDDAV_H_
#define _MOD_CARDDAV_H_

#define NS_CARDDAV "urn:ietf:params:xml:ns:carddav"

#define XML_VERSION "1.0"

#define NODE_NS(node, ns_string) (node->ns && node->ns->href && \
				  strcmp((char *) node->ns->href, ns_string) == 0)

#define FOR_CHILD(node, parent) \
	for (node = parent ? parent->children : NULL; node; node = node->next)

#define NODE_NOT_DAV(node) node->type != XML_ELEMENT_NODE || !NODE_NS(node, NS_DAV)
#define NODE_NOT_CARDDAV(node) node->type != XML_ELEMENT_NODE || !NODE_NS(node, NS_CARDDAV)

#define NODE_MATCH(node, y) (strcmp((char *) node->name, y) == 0)

typedef struct carddav_search_s carddav_search_t;

#endif /* _MOD_CARDDAV_H_ */


carddav_vcard.h
/**
 * This is part of a mod_carddav library.
 * 
 * Copyright (C) 2008 Nokia Corporation.
 * 
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */

#ifndef _CARDDAV_VCARD_H_
#define _CARDDAV_VCARD_H_

int carddav_vcard_search(const char *file, xmlNode *data, xmlNode *filter,
                         carddav_search_t **p);

char *carddav_vcard_dump(carddav_search_t *p);
int carddav_vcard_free(carddav_search_t *p);

#endif  /* _CARDDAV_VCARD_H_ */


carddav_vcard.c
/**
 * This is part of a mod_carddav library.
 *
 * Copyright (C) 2008 Nokia Corporation.
 *
 * Contact: Jari Urpalainen <jari.urpalainen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <limits.h>

#include <glib.h>
#include <libebook/e-vcard.h>

#include <libxml/tree.h>
#include "mod_carddav.h"

/** parse a vcard file */
static EVCard *parse_vcard(const char *file)
{
    EVCard *c = NULL;
    char *str;
    gsize length;

    if (g_file_get_contents(file, &str, &length, NULL)) {
	c = e_vcard_new_from_string(str);
	g_free(str);
    }

    return c;
}

static char *endswith(const char *s1, const char *s2,
                      char * (*fn)(const char *a, const char *b))
{
    char *p;

    if ((p = fn(s1, s2)) && (strlen(p) == strlen(s2)))
	return p;
    else
	return NULL;
}

static char *beginswith(const char *s1, const char *s2,
                        char * (*fn)(const char *a, const char *b))
{
    char *p;

    if ((p = fn(s1, s2)) && (p == s1))
	return p;
    else
	return NULL;
}

static char *contains(const char *s1, const char *s2,
                      char * (*fn)(const char *a, const char *b))
{
    return fn(s1, s2);
}

typedef char *fn_s(const char *a, const char *b);

/** text match checkings */
static int text_matches(xmlNode *node, const char *pch)
{
    xmlChar *collation = xmlGetProp(node, (const xmlChar *) "collation");
    xmlChar *invert = xmlGetProp(node, (const xmlChar *) "negate-condition");
    xmlChar *match_type = xmlGetProp(node, (const xmlChar *) "match-type");
    xmlChar *value = xmlNodeGetContent(node);
    int rc = 0, feq = FALSE;
    fn_s *fn;
    int (*eq) (const xmlChar *a, const xmlChar *b);
    char * (*cmp) (const char *a, const char *b, fn_s *fn) = NULL;

    if (match_type == NULL || strcasecmp((char *) match_type, "contains") == 0)
	cmp = contains;
    else  if (strcasecmp((char *) match_type, "equals") == 0)
	feq = TRUE;
    else if (strcasecmp((char *) match_type, "starts-with") == 0)
	cmp = beginswith;
    else if (strcasecmp((char *) match_type, "ends-with") == 0)
	cmp = endswith;

    if (collation == NULL ||
	strcasecmp((char *) collation, "i;unicode-casemap") == 0 ||
		strcasecmp((char *) collation, "i;ascii-casemap") == 0) {
	fn = (fn_s *) xmlStrcasestr;
	eq = xmlStrcasecmp;
    }
    else {
	fn = (fn_s *) xmlStrstr;
	eq = xmlStrcmp;
    }

    if (feq)
	rc = (value && pch) ? eq((xmlChar *) pch, value) == 0 : 0;
    else if (cmp)
	rc = (value && pch) ? cmp(pch, (char *) value, fn) != NULL : 0;

    if ((feq || cmp) && invert && strcasecmp((char *) invert, "yes") == 0)
	rc = !rc;

    xmlFree(value);
    xmlFree(match_type);
    xmlFree(invert);
    xmlFree(collation);

    return rc;
}

/** parameter tests */
static int param_matches(xmlNode *node, EVCardAttribute *attr)
{
    xmlChar *pch = xmlGetProp(node, (const xmlChar *) "name");
    GList *p;

    if (pch == NULL)
	return FALSE;

    p = e_vcard_attribute_get_params(attr);

    for ( ; p; p = p->next) {
	EVCardAttributeParam *param = p->data;
	const char *name = e_vcard_attribute_param_get_name(param);

	if (name && strcasecmp(name, (char *) pch) == 0) {
	    xmlNode *prop;

	    FOR_CHILD(prop, node) {
		if (NODE_NOT_CARDDAV(prop)) {
		    ;
		}
		else if (NODE_MATCH(prop, "is-not-defined")) {
		    xmlFree(pch);
		    return FALSE;
		}
		else if (NODE_MATCH(prop, "text-match")) {
		    GList *v = e_vcard_attribute_param_get_values(param);

		    for ( ; v; v = v->next) {
			if (text_matches(prop, v->data)) {
			    xmlFree(pch);
			    return TRUE;
			}
		    }
		}
		else if (prop->type == XML_ELEMENT_NODE) {
		    /* rule not understood !!! */
		    xmlFree(pch);
		    return FALSE;
		}
	    }
	}
    }
    xmlFree(pch);

    return FALSE;
}

static int attr_match(xmlNode *node, EVCardAttribute *attr)
{
    xmlNode *param;
    int ret = TRUE;
    xmlChar *allof = xmlGetProp(node, (const xmlChar *) "test");
    int and = (allof && strcasecmp((char *) allof, "allof") == 0);

    xmlFree(allof), allof = NULL;

    FOR_CHILD(param, node) {
	int rc = TRUE;

	if (NODE_NOT_CARDDAV(param)) {
	    if (param->type == XML_ELEMENT_NODE)
		rc = FALSE;
	    else
		continue;
	}
	else if (NODE_MATCH(param, "is-not-defined")) {
	    rc = (attr == NULL);
	}
	else if (NODE_MATCH(param, "param-filter")) {
	    rc = (attr && param_matches(param, attr)) == TRUE;
	}
	else if (NODE_MATCH(param, "text-match")) {
	    GList *v = attr ? e_vcard_attribute_get_values(attr) : NULL;

	    for (rc = FALSE; v; v = v->next) {
		if (text_matches(param, v->data)) {
		    rc = TRUE;
		    break;
		}
	    }
	}
	else if (param->type == XML_ELEMENT_NODE) {
	    rc = FALSE;
	}

	if ((and && rc == FALSE) || (and == FALSE && rc))
	    return rc;

	ret = rc;
    }

    return ret;
}

/**
 * vcard search based on <filter> rules
 */
static int carddav_vcard_matches(EVCard *c, xmlNode *filter)
{
    xmlNode *node;
    int ret = TRUE, and;
    xmlChar *allof = xmlGetProp(filter, (const xmlChar *) "test");

    and = (allof && strcasecmp((char *) allof, "allof") == 0);
    xmlFree(allof), allof = NULL;

    for (node = filter ? filter->children : NULL; node; node = node->next) {
	if (NODE_NOT_CARDDAV(node)) {
	    if (node->type == XML_ELEMENT_NODE && and)
		return FALSE;
	    else
		continue;
	}
	else if (NODE_MATCH(node, "prop-filter")) {
	    xmlChar *pch = xmlGetProp(node, (const xmlChar *) "name");
	    int rc = FALSE;
	    GList *l;

	    if (pch == NULL)
		return FALSE;

	    for (l = e_vcard_get_attributes(c); l; l = l->next) {
		EVCardAttribute *attr = l->data;
		const char *name = e_vcard_attribute_get_name(attr);

		if (name && strcasecmp((char *) pch, name) == 0) {
		    rc = attr_match(node, attr);

		    if (rc == TRUE)
			break;
		}
	    }
	    xmlFree(pch);

	    if ((and && rc == FALSE) || (and == FALSE && rc))
		return rc;

	    ret = rc;
	}
	else if (node->type == XML_ELEMENT_NODE) {
	    /* rule not understood !!! */
	    if (and)
		return FALSE;
	}
    }

    return ret;
}

/* dump VCard contents after filtering */
static char *carddav_vcard_dump_query(EVCard *c, xmlNode *carddata)
{
    int all_props = -1;
    xmlNode *n = NULL;

    FOR_CHILD(n, carddata) {
	if (NODE_NOT_CARDDAV(n))
	    ;
	else if (NODE_MATCH(n, "prop"))
	    all_props = FALSE;
	else if (NODE_MATCH(n, "allprop"))
	    all_props = TRUE;
    }

    if (all_props == FALSE) {
	GList *l = e_vcard_get_attributes(c);

	for ( ; l; ) {
	    EVCardAttribute *attr = l->data;
	    int f = FALSE;

	    FOR_CHILD(n, carddata) {
		if (NODE_NOT_CARDDAV(n)) {
		    ;
		}
		else if (NODE_MATCH(n, "prop")) {
		    xmlChar *pch = xmlGetProp(n, (const xmlChar *) "name");
		    const char *name = e_vcard_attribute_get_name(attr);

		    if (pch && name && strcasecmp((char *) pch, name) == 0) {
			f = TRUE;
			break;
		    }
		    xmlFree(pch);
		}
	    }

	    if (f == FALSE) {
		GList *t = l->next;

		e_vcard_remove_attribute(c, attr);
		l = t;
	    }
	    else {
		xmlChar *val = xmlGetProp(n, (const xmlChar *) "novalue");

		if (val && strcasecmp((char *) val, "yes") == 0) {
		    e_vcard_attribute_remove_params(attr);
		    e_vcard_attribute_remove_values(attr);
		}
		xmlFree(val);

		l = l->next;
	    }
	}
    }

    return e_vcard_to_string(c, EVC_FORMAT_VCARD_30);
}

/** local search structure */
struct carddav_search_s
{
    xmlNode *addressdata, *filter;
    EVCard *root;
};

/**
 * check if search criterias fit with this resource
 * allows request handler to omit this resource
 */
int carddav_vcard_search(const char *file, xmlNode *addressdata,
                         xmlNode *filter, carddav_search_t **pp)
{
    carddav_search_t *p = *pp = calloc(1, sizeof (*p));

    p->root = parse_vcard(file);
    p->addressdata = addressdata;
    p->filter = filter;

    if (filter == NULL)
	return 1;

    return carddav_vcard_matches(p->root, filter);
}

/** return vcard stuff for all requested reports */
char *carddav_vcard_dump(carddav_search_t *p)
{
    if (p == NULL)
	return NULL;

    /* dump full content */
    if (p->filter == NULL)
	return p->root ? e_vcard_to_string(p->root, EVC_FORMAT_VCARD_30) : NULL;

    /* dump selected components only */
    return carddav_vcard_dump_query(p->root, p->addressdata);
}

/** free resources for search/query */
int carddav_vcard_free(carddav_search_t *p)
{
    if (p == NULL)
	return -1;

    if (p->root) {
	g_object_unref(p->root);
	p->root = NULL;
    }

    free(p);

    return 0;
}
